### PR TITLE
wallet cli: donate command

### PR DESCRIPTION
**Pull request by kenshi84**
Issue #1447

### CODE REVIEW

**Comment by NanoAkron** [comment]
Does this run through always-confirm-transactions to catch accidental large value donations? Maybe explicitly confirm with the user that this is going to the donation address?

I can just see that one day where somebody says 'I accidentally donated all my monero, plz refund' - know what I mean?

**Comment by kenshi84** [comment]
Yes, it's just a simple wrapper on top of the normal `transfer` command, so the same confirmation scheme applies. See the code.

**Comment by NanoAkron** [comment]
Ah cool - thanks!

When it does come up with `confirm transfer to 44AFFq5kSiGBoZ4NMDwYtN18obc8AemS33DBLWs3H7otXft3XjrpDtQGv7SqSsaBYBb98uNbr2VBBEt7f2wfn3RVGQBEP3A`, would it be at all possible to have that address already encoded in the address book and resolve in plaintext as `Monero Donation Address` instead of the long alphanumerics, maybe to give the user some added confidence that it's really a donation?

File: src/simplewallet/simplewallet.cpp [hunk 1]
```diff
@@ -2826,6 +2827,45 @@ bool simple_wallet::sweep_all(const std::vector<std::string> &args_)
...
+bool simple_wallet::donate(const std::vector<std::string> &args_)
+{
+  std::vector<std::string> local_args = args_;
+  if(local_args.empty() || local_args.size() > 3)
+  {
+     fail_msg_writer() << tr("wrong number of arguments");
+     return true;
+  }
+  std::string mixin_str;
+  std::string address_str = "44AFFq5kSiGBoZ4NMDwYtN18obc8AemS33DBLWs3H7otXft3XjrpDtQGv7SqSsaBYBb98uNbr2VBBEt7f2wfn3RVGQBEP3A";
```
Comment (anonimal) [inline src/simplewallet/simplewallet.cpp]: I'd recommend `const` here because why not.

**Comment by xmr-eric** [comment]
Or perhaps something like `44AFFq5kSiGBoZ4NMDwYtN18obc8AemS33DBLWs3H7otXft3XjrpDtQGv7SqSsaBYBb98uNbr2VBBEt7f2wfn3RVGQBEP3A  (Monero donation address)`

**Comment by NanoAkron** [comment]
@Jaqueeee is it possible to pre-populate the address book with this donation address for everyone?

File: src/simplewallet/simplewallet.cpp [hunk 2]
```diff
@@ -663,6 +663,7 @@ simple_wallet::simple_wallet()
   m_cmd_binder.set_handler("locked_transfer", boost::bind(&simple_wallet::locked_transfer, this, _1), tr("locked_transfer [<mixin_count>] <addr> <amount> <lockblocks>(Number of blocks to lock the transaction for, max 1000000) [<payment_id>]"));
   m_cmd_binder.set_handler("sweep_unmixable", boost::bind(&simple_wallet::sweep_unmixable, this, _1), tr("Send all unmixable outputs to yourself with mixin 0"));
   m_cmd_binder.set_handler("sweep_all", boost::bind(&simple_wallet::sweep_all, this, _1), tr("sweep_all [mixin] address [payment_id] - Send all unlocked balance an address"));
+  m_cmd_binder.set_handler("donate", boost::bind(&simple_wallet::donate, this, _1), tr("donate [<mixin_count>] <amount> [payment_id] - Donate <amount> to the dev (donate.getmonero.org)."));
```
Comment (NanoAkron) [inline src/simplewallet/simplewallet.cpp]: Can you make this read 'development team' rather than the singular 'dev'?

**Comment by Jaqueeee** [comment]
@NanoAkron sure. Can even make it non deletable. ;)

**Comment by NanoAkron** [comment]
You're bound to get at least 1 complaint if you do that.

And of course, if you don't you'll get someone complaining that they accidentally deleted it ;)

**Comment by kenshi84** [comment]
Now it uses donate.getmonero.org instead of the long address.

**Comment by anonimal** [comment]
>donate.getmonero.org

I can't use this feature with the default cli wallet now. I don't want to use the gui most of the time either. What about others like me?

**Comment by fluffypony** [comment]
@anonimal can't use it because the DNS resolution fails, or some other reason?

**Comment by anonimal** [comment]
Not hardcoding the address also opens up potential attack vectors via the address book.

**Comment by anonimal** [comment]
@fluffypony `Error: wrong address: donate.getmonero.org` <-- when used via default cli (at least for me).

**Comment by fluffypony** [comment]
@anonimal that's...weird...are you just doing ```transfer donate.getmonero.org 1```?

**Comment by anonimal** [comment]
Nope, `donate 1`. This is because the hardcoded address is now `donate.getmonero.org` instead of the actual address (is there supposed to be address translation somewhere?).  #1447 was intended to make it extremely easy to donate but this donate command is presently useless unless there's an unintended bug that I don't know of which prevents it from actually donating.

**Comment by kenshi84** [comment]
@anonimal Actually I also encountered the same problem before when using a choppy internet connection. Even with a stable connection, it often takes a while to resolve the real address from OpenAlias. I suppose it's due to some DNS related issues. `donate 1` is literally equivalent to `transfer donate.getmonero.org 1`, so if you see the same error with this command, it's not an issue of this PR.

I initially used the hard coded address, but the OpenAlias version prints more helpful information in the CLI (i.e. donate.getmonero.org is translated to the real long address with the registered name 'Monero Address').

**Comment by luigi1111** [comment]
Yeah, donate looks nicer. Can still hardcode it the further down the line if it was really wanted I guess; not sure it would be or should be.

**Comment by anonimal** [comment]
@kenshi84 @luigi1111 if OpenAlias requires that #1114 be resolved, then no; this command (and I'm assuming OpenAlias (or related features)) will not work over Tor. Since I've never used aliases for tx's, I've never had a problem but, for me and other adept Tor users, I think it would be much simpler to remove the overhead and simply hardcode the address (and would also be much appreciated).

**Comment by luigi1111** [comment]
Please use #1447 for further discussion.

**Pull request merged**